dev-docs/RFCs/v8.0/luma-refactor.md

```diff
@@ -0,0 +1,83 @@
+- 14 separate modules with no clear sense of how they're meant to interoperate
+- 201 exported classes and functions, again without a clear sense of their relationships
+- A large code base. ~733kB of source files - compare to e.g. a full 3D engine like three.js at ~1.2MB (1.6x), a low-level library like PicoGL at ~183kB (0.25x)
+- Several unfinished or unused APIs that make it hard to tell what's reliable
+- Documentation is sprawling and not focused on the API. Examples demonstrate random scenes rather than API features
+
+Clearly defining luma.gl's goals as a framework, and structuring the codebase and documentation around those goals, would benefit it in the following ways:
+- Reduce its complexity and associated maintenance costs
+- Reduce the weight it imposes on applications that use it
+- Make its value proposition clear to potential users
+- Simplify adoption
+
+## Overview
+
+The first step towards simplifying luma.gl is to be explicit about what its goals are. I propose that changes be made around the following key principles:
+- Targeting intermediate to advanced WebGL users
+- All systems implemented should solve explicitly identified problems. Experiments are encouraged, but those that fail to find a use case should be removed
+- Structured around levels of abstraction, primarily helper tools for low-level WebGL work, WebGL wrapper classes, and higher-level engine features
+- Increased focus on architectural fundamentals, simplicity, building systems that are easy to reason about. Less focus on using the latest APIs or frameworks
+
+
+The actual refactor of luma.gl would occur in three phases:
+
+### Clean Out Unused Code
+
+This will warrant some discussion. The following are systems I believe should be considered for removal or consolidation:
+- `SceneGraphNode`, `BaseModel`, `Model` - functionality could be supported with just the `Model` and `Group` classes
+- `CameraNode` - Appears unused
+- Uniform animations - Complex and unused
+- `ShaderCache` - Superseded by `ProgramManager`
+- core/lighting, core/materials - currently empty or simple data classes
+- core/multipass - Appears unused
+- seer integration - Complex and will interfere with other parts of the system if enabled (e.g. GPU timing). Is anyone using it?
+- debug/parameter-definitions - Appears unused
+- main module - Appears unused
+- core/geometry - Perhaps better suited to math.gl?
+
+### Structure Modules Around Meaningful Themes
+
+I propose reducing the number of modules to five, around the following themes:
+
+#### Constants
+
+The `constants` module would remain as it is now.
+
+#### Low-level
+
+luma.gl would would contain two low-level modules:
+- `shader-tools`: Current ShaderTools and shader debugging tools from current debug module
+- `context-tools`: Current webgl-polyfill and webgl-state-tracker, and context debugging tools from current debug module
```

Comment:
These are targeting very different use cases. I am supportive of merging them into one module, but the endpoints need to be carefully designed for effective tree shaking.

Reply:
Also not against putting them in one module, but they should be kept separate code wise (not sharing code etc).

These have been kept separate for a variety:
- partly because they are very complex and different in nature
- maintainers will be tempted to create problematic "shortcuts" between them if they are in the same module.
- The polyfills were being groomed for contribution back to Khronos, so we put a lot of work into making sure they were independent.
- The state tracker was proposed to mapbox to enable them to share context with us, and we wanted it to be independent to they would not have issues with using it.
- mapbox didn't use it but we see other companies doing similar integrations, so it may yet come in handy.

Reply:
Note that these modules are quite small. The dist.min.js files are 30kB for state-tracker and 21kB for polyfill (compare to 302kB for core, 145kb for webgl, 97kB for shader-tools).